src/aws-util/src/s3_uploader.rs

```diff
@@ -0,0 +1,554 @@
+        }
+        if file_size_limit > AWS_S3_MAX_OBJECT_SIZE.as_u64() {
+            return Err(anyhow!(format!(
+                "invalid file size: {}, cannot exceed {} bytes",
+                file_size_limit,
+                AWS_S3_MAX_OBJECT_SIZE.as_u64()
+            )));
+        }
+        let max_parts_count: u64 = AWS_S3_MAX_PART_COUNT.try_into().expect("i32 to u64");
+        if file_size_limit / part_size_limit > max_parts_count {
+            return Err(anyhow!(format!(
+                "total number of possible parts (file_size_limit / part_size_limit): {}, cannot exceed {}",
+                file_size_limit / part_size_limit,
+                AWS_S3_MAX_PART_COUNT
+            )));
+        }
+        Ok(S3MultiPartUploaderConfig {
+            part_size_limit,
+            file_size_limit,
+        })
+    }
+
+    fn file_size_limit(&self) -> u64 {
+        self.file_size_limit
+    }
+
+    fn part_size_limit(&self) -> u64 {
+        self.part_size_limit
+    }
+}
+
+impl Default for S3MultiPartUploaderConfig {
+    fn default() -> Self {
+        Self {
+            part_size_limit: Self::DEFAULT_PART_SIZE_LIMIT.as_u64(),
+            file_size_limit: Self::DEFAULT_MAX_FILE_SIZE.as_u64(),
+        }
+    }
+}
+
+impl S3MultiPartUploader {
+    /// Creates a an instance of `S3MultiPartUploader` for the given `bucket` and `path`.
+    /// This starts the multi part upload by making a `create_multipart_upload` call, and
+    /// initializes all the internal state required to track the ongoing upload.
+    pub async fn try_new(
+        sdk_config: &SdkConfig,
+        bucket: String,
+        path: String,
+        config: S3MultiPartUploaderConfig,
+    ) -> Result<S3MultiPartUploader, S3MultiPartUploadError> {
+        let client = crate::s3::new_client(sdk_config);
+        let res = client
+            .create_multipart_upload()
+            .bucket(&bucket)
+            .key(&path)
+            .send()
+            .await?;
+        let upload_id = res
+            .upload_id()
+            .ok_or(anyhow!(
+                "create_multipart_upload response missing upload id"
+            ))?
+            .to_string();
+        Ok(S3MultiPartUploader {
+            client,
+            bucket,
+            path,
+            upload_id,
+            part_count: 0,
+            etags: Default::default(),
+            total_bytes_uploaded: 0,
+            buffer: Default::default(),
+            config,
+        })
+    }
+
+    /// Adds the `data` to the internal buffer and flushes the buffer if it is more than
+    /// the part threshold defined in `S3MultiPartUploaderConfig`.
+    /// Note, the size of the data cannot exceed the possible total size of one single file,
+    /// returned by `config.file_size_limit()`.
+    pub async fn add_chunk(
+        &mut self,
+        data: Bytes,
+    ) -> Result<InProgressUpload, S3MultiPartUploadError> {
+        let num_of_bytes: u64 = data.len().try_into().expect("usize to u64");
```

Comment:
We force the use of 64-bit platforms, so you can use `ore::cast::CastFrom` here to avoid the panic.